benchmarks/cpp/p2p_communication.cpp

```diff
@@ -0,0 +1,172 @@
+  const DeviceIdxType size = communicator->size();
+  const DeviceIdxType send_peer = (my_rank + 1) % size;
+  const DeviceIdxType recv_peer = (size + my_rank - 1) % size;
+
+  if (my_rank == 0) {
+    std::cout << "Starting P2P communication benchmark..." << std::endl;
+    std::cout << "Repetitions per size: " << kNumRepetitions << std::endl;
+    std::cout << "Number of devices: " << size << std::endl;
+    std::cout << "Testing tensor sizes from 2^10 to 2^26 elements" << std::endl;
+    std::cout << std::endl;
+
+    // Table header
+    std::cout << std::left << std::setw(15) << "Message Size" << std::setw(12)
+              << "Elements" << std::setw(15) << "Latency (μs)" << std::setw(18)
+              << "Bandwidth (GB/s)" << std::endl;
+    std::cout << std::string(60, '-') << std::endl;
+  }
+
+  // Test each tensor size
+  for (size_t size_idx = 0; size_idx < tensor_sizes.size(); size_idx++) {
+    const int current_tensor_size = tensor_sizes[size_idx];
+
+    auto container = std::make_unique<hir::HostIrContainer>();
+    FusionGuard fg(container.get());
+
+    // Create the P2P communication setup
+    auto* send_peer_val = IrBuilder::create<Val>(send_peer, DataType::Int);
+    auto* recv_peer_val = IrBuilder::create<Val>(recv_peer, DataType::Int);
+
+    auto* send_tv = makeContigTensor(1);
+    auto* recv_tv = makeContigTensor(1);
+    container->addInput(send_tv);
+    container->addInput(recv_tv);
+
+    auto send = IrBuilder::create<P2PCommunication>(
+        P2PCommunicationType::SEND,
+        send_tv,
+        send_peer_val,
+        CommunicatorBackend::kCuda);
+    auto recv = IrBuilder::create<P2PCommunication>(
+        P2PCommunicationType::RECV,
+        recv_tv,
+        recv_peer_val,
+        CommunicatorBackend::kCuda);
+
+    std::vector<P2PCommunication*> grouped_communications = {send, recv};
+    auto share_mem_handles = IrBuilder::create<hir::ShareMemHandles>(
+        std::move(grouped_communications));
+    auto wait_send = IrBuilder::create<hir::Wait>(send);
+    auto wait_recv = IrBuilder::create<hir::Wait>(recv);
+
+    container->pushBackTopLevelExprs(share_mem_handles);
+    container->pushBackTopLevelExprs(send);
+    container->pushBackTopLevelExprs(recv);
+    container->pushBackTopLevelExprs(wait_send);
+    container->pushBackTopLevelExprs(wait_recv);
+
+    hir::HostIrEvaluator executor(std::move(container), communicator);
+
+    // Create tensors
+    at::TensorOptions tensor_options =
+        at::TensorOptions().device(at::kCUDA, my_rank);
+    at::Tensor send_tensor = at::empty({current_tensor_size}, tensor_options);
+    at::Tensor recv_tensor = at::empty({current_tensor_size}, tensor_options);
+
+    std::unordered_map<Val*, PolymorphicValue> inputs = {
+        {send_tv, send_tensor}, {recv_tv, recv_tensor}};
+
+    // Calculate data transfer size
+    double data_size_mb =
+        (current_tensor_size * sizeof(float)) / (1024.0 * 1024.0);
```

Comment:
Replacing `float` with `dtype` derived from tensor options or tensors is more error-proof and extensible.
Same for other instances.